docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1164 @@
+- Error condition
+  - if *addr* is correct: precondition violated. [FS-A-COMM]
+  - if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+----
+
+### Temporal Properties
+
+
+#### Fairness
+
+As mentioned above, without assumptions on the correctness of some
+peers, no protocol can achieve the required specifications. Therefore, 
+we sometimes consider the following (fairness) constraint in the
+safety and liveness properties below:
+
+
+#### **[FS-SOME-CORR-PEER]**:
+Initially, the set *peerIDs* contains at least one correct full node.
+
+> While in principle the above condition can be part of a sufficient
+> condition to solve [FS-SEQ-LIVE] and
+> [FS-SEQ-TERM] and [FS-SEQ-TERM-SYNC], we will see below that the
+> current implementation of Fastsync (V2) requires the following (much
+> stronger) requirement
+
+#### **[FS-ALL-CORR-PEER]**:
+At all times, the set *peerIDs* contains only correct full nodes.
+
+
+#### Safety
+
+
+<!--
+> safety specifications / invariants in English 
+---->
+
+
+#### **[FS-VC-ALL-CORR-NONABORT]**:
+Under [FS-ALL-CORR-PEER], *Fastsync* never terminates with failure.
+
+
+#### **[FS-VC-STATE-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
+application state is the one that corresponds to the blockchain at
+height *terminationHeight*.
+
+#### **[FS-VC-BLOCKS-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
+returned list of blocks  is the one that corresponds to the blocks of
+the
+blockchain.
+
+
+> As this specification does
+> not assume that a correct peer is at the most recent height
+> of the blockchain (it might lag behind), the property [FS-SEQ-TERM]
+> cannot be ensured in an unreliable distributed setting. We consider
+> the following relaxation. (Which is typically sufficient for
+> Tendermint, as the consensus reactor then synchronizes from that
+> height.)
+
+#### **[FS-VC-CORR-INV]**:
+Under [FS-ALL-CORR-PEER], let *maxh* be the maximum 
+height of a correct peer [**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link]
+in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates
+successfully, it is at some height *terminationHeight >= maxh - 1*.
+
+> The above property is independent of how many blocks are added to the
+> blockchain (and learned by the peers) while *Fastsync* is running. It
+> links the target height to the initial state. If *Fastsync* has to
+> catch-up many blocks, it would be better to  link the
+> target height to a time close to the termination. This is captured by
+> the following specification:
+
+
+#### **[FS-VC-CORR-INV-SYNC]**:
+Under [FS-ALL-CORR-PEER], there exists a constant time interval *TD*, such
+that if *term* is the time *Fastsync* terminates and
+*maxh* be the maximum height of a correct peer
+[**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link] in *peerIDs* at the time
+*term - TD*, then if *FastSync* terminates successfully, it is at
```

Comment:
Sequential property (FS-SEQ-TERM-SYNC) refers to a logical diff (parameter D that is expressed as lag in number of heights), while here we introduce something similar but not the same with time parameter TD. Can we somehow align this better or it is fine as sequential spec is defined without reference to system parameters.

Reply:
I agree the issue here is not explicitly discussed. The problem is that in the case the peers are themselves synchronized to the blockchain (they are at the current height), then *TD* together with LTIME in [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]  can be used to bound $D$. If the peers themselves lag behind and catch-up during the *TD* time, we cannon bound anything. Perhaps it is enough to add an explanation at this point to make it clear. What do you think?